codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessageEncoder.java

```diff
@@ -0,0 +1,112 @@
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandler.Sharable;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.MessageToByteEncoder;
+import io.netty.handler.codec.haproxy.HAProxyProxiedProtocol.AddressFamily;
+import io.netty.util.CharsetUtil;
+import io.netty.util.NetUtil;
+
+import java.util.List;
+
+import static io.netty.handler.codec.haproxy.HAProxyConstants.*;
+
+/**
+ * Encodes an HAProxy proxy protocol message
+ *
+ * @see <a href="http://www.haproxy.org/download/1.8/doc/proxy-protocol.txt">Proxy Protocol Specification</a>
+ */
+@Sharable
+public class HAProxyMessageEncoder extends MessageToByteEncoder<HAProxyMessage> {
+
+    private static final int V2_VERSION_BITMASK = 0x02 << 4;
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, HAProxyMessage msg, ByteBuf out) throws Exception {
+        if (msg.protocolVersion() == HAProxyProtocolVersion.V1) {
+            encodeV1(msg, out);
+        } else if (msg.protocolVersion() == HAProxyProtocolVersion.V2) {
+            encodeV2(msg, out);
+        } else {
+            throw new HAProxyProtocolException("Unsupported version: " + msg.protocolVersion());
+        }
+    }
+
+    private static void encodeV1(HAProxyMessage msg, ByteBuf out) {
+        final String protocol = msg.proxiedProtocol().name();
+        StringBuilder sb = new StringBuilder(108)
+                .append("PROXY ").append(protocol).append(' ')
+                .append(msg.sourceAddress()).append(' ')
+                .append(msg.destinationAddress()).append(' ')
+                .append(msg.sourcePort()).append(' ').append(msg.destinationPort()).append("\r\n");
+        out.writeBytes(sb.toString().getBytes(CharsetUtil.US_ASCII));
+    }
+
+    private static void encodeV2(HAProxyMessage msg, ByteBuf out) {
+        out.writeBytes(BINARY_PREFIX);
+        out.writeByte(V2_VERSION_BITMASK | msg.command().byteValue());
+        out.writeByte(msg.proxiedProtocol().byteValue());
+
+        if (msg.proxiedProtocol().addressFamily() == AddressFamily.AF_IPv4 ||
+            msg.proxiedProtocol().addressFamily() == AddressFamily.AF_IPv6) {
+            byte[] sourceAddress = NetUtil.createByteArrayFromIpAddressString(msg.sourceAddress());
+            byte[] destinationAddress = NetUtil.createByteArrayFromIpAddressString(msg.destinationAddress());
+            out.writeShort(sourceAddress.length + destinationAddress.length + 4 + msg.tlvNumBytes());
+            out.writeBytes(sourceAddress);
+            out.writeBytes(destinationAddress);
+            out.writeShort(msg.sourcePort());
+            out.writeShort(msg.destinationPort());
+            encodeTlvs(msg.tlvs(), out);
+        } else if (msg.proxiedProtocol().addressFamily() == AddressFamily.AF_UNIX) {
+            out.writeShort(216 + msg.tlvNumBytes());
+            byte[] srcAddressBytes = msg.sourceAddress().getBytes(CharsetUtil.US_ASCII);
+            out.writeBytes(srcAddressBytes);
```

Comment:
again consider using `out.writeCharSequence(...)`

Reply:
I think we have a couple of options:
1) Pad the string to 108 length, and use `writeCharSequence `
    -> I'm not sure the extra string allocation will be worth it. I don't think readability will improve either, since we still need to 1) get the size of the address in bytes 2) allocate a new 108 byte length string 3) and write to `ByteBuffer`.
2) Restrict user input to accept only 108 byte addresses
    -> usability 😢 
3) Keep AS-IS

I prefer 3, but might be willing to change to 1) if you wish